src/vm/analysis/trait_checker/pre_type_check.rs

```diff
@@ -0,0 +1,176 @@
+use std::collections::{HashMap, HashSet};
+
+use vm::analysis::types::{ContractAnalysis, AnalysisPass};
+use vm::analysis::AnalysisDatabase;
+use vm::analysis::errors::{CheckResult, CheckError, CheckErrors};
+use vm::representations::{SymbolicExpression, ClarityName};
+use vm::representations::SymbolicExpressionType::{AtomValue, Atom, List, LiteralValue, TraitReference, Field};
+use vm::types::{Value, TraitIdentifier};
+use vm::functions::NativeFunctions;
+use vm::functions::define::{DefineFunctions, DefineFunctionsParsed};
+
+pub struct PreTypeCheckingTraitChecker {
+}
+
+impl AnalysisPass for PreTypeCheckingTraitChecker {
+
+    fn run_pass(contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+        let mut command = PreTypeCheckingTraitChecker::new();
+        command.run(contract_analysis, analysis_db)?;
+        Ok(())
+    }
+}
+
+impl PreTypeCheckingTraitChecker {
+
+    fn new() -> Self {
+        Self {
+        }
+    }
+
+    pub fn run(&mut self, contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+        let exprs = contract_analysis.expressions[..].to_vec();
+        let mut trait_usages = self.find_trait_usages(&exprs)?;
+        
+        // Presence of orphaned traits should throw
+        for (orphan, expr) in trait_usages.orphan_trait_references {
```

Comment:
What happens if there are multiple orphan trait references?  Is this code supposed to silently drop all other errors except for the first one?

Reply:
Correct, this is the current pattern adopted: when the analysis is failing, we abort it and return the first error (and diagnostic) encountered.

Reply:
Okay, sounds good :+1: